### PR TITLE
Adding the library screen

### PR DESCRIPTION
**Pull request by Kenn203**

### CODE REVIEW

File: app/src/main/res/layout/activity_search.xml [hunk 1]
```diff
@@ -1,52 +1,78 @@
...
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_alignParentTop="true"
+                android:layout_centerHorizontal="true"
+                android:layout_marginLeft="12dp"
+                android:layout_toRightOf="@+id/ivBookCover"
+                android:text="Title"
+                android:textSize="15sp"
+                android:textStyle="bold"/>
+            <LinearLayout
```
Comment (Saneet) [inline app/src/main/res/layout/activity_search.xml:40]: Good job using LinearLayout

**Review by Saneet — APPROVED** [review]

File: app/src/main/java/com/example/bookclub/ViewActivity.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        webView.getSettings().setRenderPriority(WebSettings.RenderPriority.HIGH);
+        webView.getSettings().setAppCacheEnabled(true);
+        webView.setVerticalFadingEdgeEnabled(false);
+        webView.setHorizontalScrollBarEnabled(false);
+        settings.setDomStorageEnabled(true);
+        settings.setLayoutAlgorithm(WebSettings.LayoutAlgorithm.NARROW_COLUMNS);
+        settings.setUseWideViewPort(true);
+        settings.setSavePassword(true);
+        settings.setSaveFormData(true);
+        settings.setEnableSmoothTransition(true);
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/ViewActivity.java]: You can pull this out into a different function.

File: app/src/main/java/com/example/bookclub/ViewActivity.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+        bundle = getIntent().getExtras();
+        String idUrl = bundle.getString("id");
+        webView.loadUrl("https://openlibrary.org/books/"+ idUrl);
+
+        webView.setWebViewClient(new WebViewClient(){
+            @Override
+            public void onPageStarted(WebView view, String url, Bitmap favicon) {
+                super.onPageStarted(view, url, favicon);
+                progressBar.setVisibility(View.GONE);
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/ViewActivity.java]: Maybe set this to VISIBLE explicitly in the code.

File: app/src/main/res/layout/activity_search.xml [hunk 4]
```diff
@@ -1,52 +1,78 @@
...
+                android:id="@+id/bottom_navigation"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:layout_alignParentStart="true"
+                android:layout_alignParentEnd="true"
+                android:layout_alignParentBottom="true"
+                android:layout_marginStart="2dp"
+                android:layout_marginEnd="-1dp"
+                android:layout_marginBottom="1dp"
+                app:menu="@menu/menu_bottom_navigation" />
```
Comment (Saneet) [inline app/src/main/res/layout/activity_search.xml]: Are you sure this should be inside the cardview?

File: app/src/main/res/layout/activity_view.xml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            android:layout_alignParentStart="true"
+            android:layout_centerVertical="true"/>
+
+        <WebView
+            android:id="@+id/webViewID"
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintStart_toStartOf="parent">
```
Comment (Saneet) [inline app/src/main/res/layout/activity_view.xml]: These attributes are for ConstraintLayout.

File: app/src/main/res/layout/content_book_detail.xml [hunk 6]
```diff
@@ -2,6 +2,7 @@
 <androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
     xmlns:tools="http://schemas.android.com/tools"
+    android:id="@+id/mAddBook"
```
Comment (Saneet) [inline app/src/main/res/layout/content_book_detail.xml]: The naming convention for layout element ids is @+id/add_book.

File: app/src/main/res/layout/content_book_detail.xml [hunk 7]
```diff
@@ -81,5 +82,14 @@
             android:layout_marginTop="8dp"
             android:text="Pages N/A"
             android:textSize="14sp" />
+
+        <Button
+            android:id="@+id/button2"
```
Comment (Saneet) [inline app/src/main/res/layout/content_book_detail.xml]: We can have a readable id here.

**Review by Saneet — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        View view = LayoutInflater.from(context).inflate(R.layout.activity_library, parent, false);
+        return new ViewHolder(view);
+
+    }
+
+    @Override
+    public void onBindViewHolder(@NonNull LibraryItemAdapter.ViewHolder holder, int position) {
+        LibraryItem libraryItem = libraryItemList.get(position);
+        holder.bind(libraryItem);
+
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java]: Empty lines can be removed.

File: app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import com.bumptech.glide.Glide;
+import com.example.bookclub.R;
+import com.example.bookclub.models.LibraryItem;
+
+import java.util.List;
+
+public class LibraryItemAdapter extends RecyclerView.Adapter<LibraryItemAdapter.ViewHolder> {
+    private Context context;
+    private List<LibraryItem> libraryItemList;
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java]: We can use hungarian notations here. mContext and mLibraryItemList.

File: app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+    }
+
+    @Override
+    public int getItemCount() {return libraryItemList.size();}
+
+    public class ViewHolder extends RecyclerView.ViewHolder {
+        private TextView tvAuthor;
+        private TextView tvTitle;
+        private ImageView ivBookCover;
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java]: Hungarian notations here too.

File: app/src/main/java/com/example/bookclub/BookDetailsActivity.java [hunk 11]
```diff
@@ -49,9 +57,9 @@ public class BookDetailsActivity extends AppCompatActivity {
     private TextView mBookAuthor;
     private TextView mBookPublisher;
     private TextView mBookPages;
+    private Button mAddToLibrary;
     private FloatingActionButton shareButton;
     private FloatingActionButton viewButton;
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/BookDetailsActivity.java]: Missing hungarian notation here.

File: app/src/main/java/com/example/bookclub/BookDetailsActivity.java [hunk 12]
```diff
@@ -61,12 +69,34 @@ protected void onCreate(Bundle savedInstanceState) {
...
+                LibraryItem libraryItem = new LibraryItem(book);
+                libraryItem.saveInBackground(new SaveCallback() {
+                    @Override
+                    public void done(ParseException e) {
+                        if (e != null){
+                            Toast.makeText(BookDetailsActivity.this, "Adding to library unsuccessful", Toast.LENGTH_SHORT).show();
+                            return;
+                        }
+                        currentUser.add("LibraryItem", libraryItem);
+                        currentUser.saveInBackground();
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/BookDetailsActivity.java:88]: Can we have error handling for this one too?

File: app/src/main/java/com/example/bookclub/BookDetailsActivity.java [hunk 13]
```diff
@@ -205,6 +235,7 @@ public boolean onOptionsItemSelected(MenuItem item) {
         return super.onOptionsItemSelected(item);
     }
 
+
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/BookDetailsActivity.java:234]: Looks like empty line.

File: app/src/main/java/com/example/bookclub/LibraryActivity.java [hunk 14]
```diff
@@ -1,14 +1,78 @@
...
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import java.util.ArrayList;
+import java.util.List;
 
 public class LibraryActivity extends AppCompatActivity {
+    private RecyclerView rvLibraryItem;
+    protected LibraryItemAdapter libraryItemAdapter;
+    protected List<LibraryItem> allLibraryItems;
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/LibraryActivity.java]: Hungarian notation.

File: app/src/main/java/com/example/bookclub/LibraryActivity.java [hunk 15]
```diff
@@ -1,14 +1,78 @@
...
+
+        rvLibraryItem = findViewById(R.id.rvLibraryItem);
+        allLibraryItems = new ArrayList<>();
+        libraryItemAdapter = new LibraryItemAdapter(this, allLibraryItems);
+        rvLibraryItem.setAdapter(libraryItemAdapter);
+        rvLibraryItem.setLayoutManager(new LinearLayoutManager(this));
+        populateLibraryItem();
+    }
+
+    private void populateLibraryItem(){
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/LibraryActivity.java]: We should call it populateLibraryItem's' as it populates multiple items.

File: app/src/main/java/com/example/bookclub/LibraryActivity.java [hunk 16]
```diff
@@ -1,14 +1,78 @@
...
+        ParseUser currentUser = ParseUser.getCurrentUser();
+        JSONArray itemsInLibrary = currentUser.getJSONArray("LibraryItem");
+        if(itemsInLibrary == null){
+            return;
+        }
+        for (int i = 0; i < itemsInLibrary.length(); i++){
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/LibraryActivity.java]: Space before '{'. You can just run the formatter and it will make the changes for you.

File: app/src/main/java/com/example/bookclub/LibraryActivity.java [hunk 17]
```diff
@@ -1,14 +1,78 @@
...
+            try {
+                JSONObject libraryItem = (JSONObject) itemsInLibrary.get(i);
+                String itemId = (String) libraryItem.get("objectId");
+                queryLibraryitems(itemId);
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/LibraryActivity.java:55]: Is it possible to query multiple items with a list of IDs in parse. This way we only need one network call.

File: app/src/main/java/com/example/bookclub/LibraryActivity.java [hunk 18]
```diff
@@ -1,14 +1,78 @@
...
+
+        query.getInBackground(itemId, new GetCallback<LibraryItem>() {
+            @Override
+            public void done(LibraryItem books, ParseException e) {
+                if(e == null){
+                    allLibraryItems.add(books);
+                    libraryItemAdapter.notifyDataSetChanged();
+                }
+                else{
+
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/LibraryActivity.java]: I think you forgot to handle this case.

File: app/src/main/java/com/example/bookclub/SearchActivity.java [hunk 19]
```diff
@@ -156,11 +157,20 @@ public boolean onOptionsItemSelected(MenuItem item) {
             Toast.makeText(SearchActivity.this, "Logout Successfully!", Toast.LENGTH_SHORT).show();
             onLogout();
             return true;
+        }else if(id == R.id.action_library){
+            Toast.makeText(SearchActivity.this, "now in library", Toast.LENGTH_SHORT).show();
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/SearchActivity.java]: This message is user visible so it can be added to strings.xml instead of having a literal.

File: app/src/main/java/com/example/bookclub/SearchActivity.java [hunk 20]
```diff
@@ -156,11 +157,20 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+            onLibrary();
+            return true;
         }
 
         return super.onOptionsItemSelected(item);
     }
 
+    private void onLibrary() {
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/SearchActivity.java]: These methods can be called onClickLibrary and onClickLogout so it is clear that they are user actions.

File: app/src/main/java/com/example/bookclub/models/LibraryItem.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+    public static final String KEY_BOOK_TITLE = "bookTitle";
+    public static final String KEY_COVER_URL = "coverUrl";
+
+    public LibraryItem(){}
+
+    public LibraryItem(Book book){
+        setAuthorName(book.getAuthor());
+        setBookTitle(book.getTitle());
+        setCoverUrl(book.getCoverUrl());
+
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/models/LibraryItem.java]: Empty line

**Review by Saneet — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 22]
```diff
@@ -50,5 +51,6 @@ dependencies {
     annotationProcessor 'org.parceler:parceler:1.1.12'
     implementation 'org.parceler:parceler-api:1.1.12'
 
+
```
Comment (vciancio) [inline app/build.gradle]: Remove Extra Line - Before you make a pull request, I'd recommend taking a look at the changes between your branch & master. Try cleaning up your branch by removing unnecessary formatting or changes (extra line). This will make it easier on the people reviewing your code as they don't have as many changes to process and you can proactively prevent lint / formatting errors 😄

File: app/src/main/AndroidManifest.xml [hunk 23]
```diff
@@ -8,19 +8,22 @@
...
+        android:networkSecurityConfig="@xml/network_security_config"
         android:roundIcon="@mipmap/ic_launcher_round"
         android:supportsRtl="true"
         android:theme="@style/Theme.BookClub"
         tools:targetApi="31">
         <activity
             android:name=".LibraryActivity"
             android:exported="false" />
+        <activity
+            android:name=".ViewActivity"
```
Comment (vciancio) [inline app/src/main/AndroidManifest.xml]: nit: Instead of `ViewActivity`, i'd recommend being a bit more descriptive, like `WebViewActivity`. View is pretty generic. Since you're the only one working on it, it's not too big of a deal as long as you know, but when working in larger code bases, it's good practice to be as descriptive as possible.

File: app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java [hunk 24]
```diff
@@ -0,0 +1,68 @@
...
+
+    public LibraryItemAdapter(Context context, List<LibraryItem> libraryItemList){
+        this.context = context;
+        this.libraryItemList = libraryItemList;
+    }
+
+    @NonNull
+    @Override
+    public LibraryItemAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
+        View view = LayoutInflater.from(context).inflate(R.layout.activity_library, parent, false);
```
Comment (vciancio) [inline app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java]: _Naming_: I'm a bit confused. You're inflating the `activity_library` layout? You should make sure to properly name your layouts. The Activity Layout's normally are `activity_xxxx`, list items could be something like `list_item_xxxx` (or `li_xxxx`, fragments as `fragment_xxxx` or a custom view as `view_xxxx`. As you add more views to your app, you want to make sure you keep things organized.

File: app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+            ivBookCover = itemView.findViewById(R.id.ivBookCover);
+        }
+
+        public void bind(LibraryItem item){
+            tvAuthor.setText(item.getAuthorName());
+            tvTitle.setText(item.getBookTitle());
+            String image = item.getCoverUrl();
+
+            if (image != null){
+                Glide.with(context).load(image).into(ivBookCover);
```
Comment (vciancio) [inline app/src/main/java/com/example/bookclub/Adapters/LibraryItemAdapter.java]: Is there a default image here?

File: app/src/main/java/com/example/bookclub/BookDetailsActivity.java [hunk 26]
```diff
@@ -180,7 +212,7 @@ private Uri getLocalBitmapUri(ImageView shareImage) {
     }
 
     public void viewIntent(final String id) {
```
Comment (vciancio) [inline app/src/main/java/com/example/bookclub/BookDetailsActivity.java]: I'd make this method a bit more descriptive, like `launchWebView` or something of the sort.

File: app/src/main/java/com/example/bookclub/SearchActivity.java [hunk 27]
```diff
@@ -156,11 +157,20 @@ public boolean onOptionsItemSelected(MenuItem item) {
             Toast.makeText(SearchActivity.this, "Logout Successfully!", Toast.LENGTH_SHORT).show();
             onLogout();
             return true;
+        }else if(id == R.id.action_library){
```
Comment (vciancio) [inline app/src/main/java/com/example/bookclub/SearchActivity.java]: don't forget the space after the `}`

~~~
} else if (...) {
~~~

File: app/src/main/java/com/example/bookclub/ViewActivity.java (hunk 2 above)
Comment (vciancio) [reply app/src/main/java/com/example/bookclub/ViewActivity.java]: You can have something like `setupWebView()` as a utility method to clean up the function and make it easier to read.

Your methods _should_ do 1 thing and the number of lines _should_ be short. Obviously you can't always but it's a good rule of thumb. https://stackoverflow.com/a/611322 summarizes some popular opinions.

File: app/src/main/java/com/example/bookclub/ViewActivity.java [hunk 28]
```diff
@@ -0,0 +1,76 @@
...
+                    if(webView.canGoBack()){
+                        webView.goBack();
+                    }else {
+                        finish();
+                    }
+                    return true;
+            }
+        }
+        return super.onKeyDown(keyCode, event);
+
```
Comment (vciancio) [inline app/src/main/java/com/example/bookclub/ViewActivity.java]: extra line

File: app/src/main/res/layout/activity_book_details.xml [hunk 29]
```diff
@@ -17,8 +17,8 @@
         android:layout_height="wrap_content"
         android:layout_gravity="bottom|end"
         android:layout_marginBottom="16dp"
-        app:fab_addButtonColorNormal="#5f4339"
-        app:fab_addButtonColorPressed="#744F43"
+        app:fab_addButtonColorNormal="#7F00FF"
```
Comment (vciancio) [inline app/src/main/res/layout/activity_book_details.xml]: You can extract these colors out to the `colors.xml` file. [Android Docs Link](https://developer.android.com/guide/topics/resources/more-resources#Color)

File: app/src/main/res/layout/activity_library.xml [hunk 30]
```diff
@@ -1,9 +1,77 @@
...
     android:layout_width="match_parent"
-    android:layout_height="match_parent"
-    tools:context=".LibraryActivity">
+    android:layout_height="wrap_content"
+    android:orientation="vertical">
 
-</androidx.constraintlayout.widget.ConstraintLayout>
\ No newline at end of file
+    <androidx.cardview.widget.CardView
+        android:id="@+id/CardViewID"
```
Comment (vciancio) [inline app/src/main/res/layout/activity_library.xml:9]: id's are normally lowercase and use [snake case](https://en.wikipedia.org/wiki/Snake_case)  (TLDR: underscores to separate words and all lowercase)

Here's a [good blog post](https://jeroenmols.com/blog/2016/03/07/resourcenaming/) on the topic of xml naming & conventions in android.

File: app/src/main/res/layout/content_book_detail.xml (hunk 6 above)
Comment (vciancio) [reply app/src/main/res/layout/content_book_detail.xml]: See [this blog post](https://jeroenmols.com/blog/2016/03/07/resourcenaming/) for xml naming conventions.

File: app/src/main/res/layout/content_book_detail.xml [hunk 31]
```diff
@@ -81,5 +82,15 @@
...
             android:textSize="14sp" />
+
+        <Button
+            android:id="@+id/mAddToLibrary"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center"
+            android:layout_marginTop="8dp"
+            android:gravity="center"
+            android:text="Add to Library" />
```
Comment (vciancio) [inline app/src/main/res/layout/content_book_detail.xml]: User facing text should be put in `strings.xml`

File: app/src/main/java/com/example/bookclub/Adapters/BookAdapter.java [hunk 32]
```diff
@@ -16,13 +20,20 @@
...
 
 import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
+import java.util.Locale;
 
 public class BookAdapter extends RecyclerView.Adapter<BookAdapter.ViewHolder> {
     private List<Book> mBooks;
     private Context mContext;
     private OnItemClickListener listener;
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/Adapters/BookAdapter.java:30]: mListener

File: app/src/main/java/com/example/bookclub/BookDetailsActivity.java [hunk 33]
```diff
@@ -69,12 +65,12 @@ protected void onCreate(Bundle savedInstanceState) {
 
         queue = Volley.newRequestQueue(this);
 
-        mAddToLibrary = findViewById(R.id.mAddToLibrary);
-        mBookImg = findViewById(R.id.mBookImg);
-        mBookTitle = findViewById(R.id.mTitleID);
-        mBookAuthor = findViewById(R.id.mAuthorID);
-        mBookPublisher = findViewById(R.id.mPublishedBy);
-        mBookPages = findViewById(R.id.mPages);
+        mAddToLibrary = findViewById(R.id.tvAddToLibrary);
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/BookDetailsActivity.java:68]: tvAddToLibrary should be tv_add_to_library

File: app/src/main/java/com/example/bookclub/SearchActivity.java [hunk 34]
```diff
@@ -92,7 +102,7 @@ public boolean onNavigationItemSelected(@NonNull MenuItem item) {
 
         rvBooks.setAdapter(bookAdapter);
         rvBooks.setLayoutManager(new LinearLayoutManager(this));
-        fetchBooks("Oscar Wilde");
+        fetchBooks("Malcolm Gladwell");
```
Comment (Saneet) [inline app/src/main/java/com/example/bookclub/SearchActivity.java:105]: Can you remove this later on? Code committed to master should be deployable code.

**Review by Saneet — CHANGES_REQUESTED** [review]
I have some comments. Let's try to land this PR today and continue the rest of the work in another Pull Request/Branch.

File: app/src/main/java/com/example/bookclub/SearchActivity.java (hunk 34 above)
Comment (Kenn203) [reply app/src/main/java/com/example/bookclub/SearchActivity.java:105]: I added that line so the user could see some books on the search screen when he on there. I didn't want to display a blank screen to the user.

Comment (Saneet) [reply app/src/main/java/com/example/bookclub/SearchActivity.java:105]: I think it's fine if the screen is blank. Or you can also pull the last search's results there so it's more relevant to the user.

**Review by Saneet — APPROVED** [review]
